quinn-proto/src/connection/mod.rs

```diff
@@ -362,6 +435,85 @@ impl Connection {
         this
     }
 
+    /// Set up qlog for this connection.
+    #[cfg(feature = "qlog")]
+    pub fn set_qlog(
+        &mut self,
+        writer: Box<dyn std::io::Write + Send + Sync>,
+        title: String,
+        description: String,
+    ) {
+        let vp = if self.side.is_server() {
+            qlog::VantagePointType::Server
+        } else {
+            qlog::VantagePointType::Client
+        };
+
+        let level = qlog::events::EventImportance::Core;
+
+        let trace = qlog::TraceSeq::new(
+            qlog::VantagePoint {
+                name: None,
+                ty: vp,
+                flow: None,
+            },
+            Some(title.to_string()),
+            Some(description.to_string()),
+            Some(qlog::Configuration {
+                time_offset: Some(0.0),
+                original_uris: None,
+            }),
+            None,
+        );
+
+        let mut streamer = qlog::streamer::QlogStreamer::new(
+            qlog::QLOG_VERSION.to_string(),
+            Some(title),
+            Some(description),
+            None,
+            std::time::Instant::now(),
+            trace,
+            level,
+            writer,
+        );
+
+        streamer.start_log().ok();
+
+        self.qlog_streamer = Some(streamer);
+    }
+
+    /// Emit a `MetricsUpdated` event to the qlog streamer containing only the updated values.
+    #[cfg(feature = "qlog")]
+    fn emit_qlog_recovery_metrics(&mut self) {
+        if self.qlog_streamer.is_none() {
```

Comment:
style: could we use let-else here rather than matching on the pattern redundantly below?